File: cpp/validation/validation.cpp
#include "autorigami/validation.h"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <limits>
#include <stdexcept>

namespace autorigami {

namespace {

[[nodiscard]] double squared_distance(const Vec3& left, const Vec3& right) {
    return norm2(left - right);
}

[[nodiscard]] std::vector<double> compute_arclengths(const std::vector<Vec3>& points) {
    std::vector<double> arclengths(points.size(), 0.0);
    for (std::size_t index = 1; index < points.size(); ++index) {
        arclengths[index] = arclengths[index - 1] + norm(points[index] - points[index - 1]);
    }
    return arclengths;
}

[[nodiscard]] std::vector<double> compute_curvature_radius(const std::vector<Vec3>& points) {
    std::vector<double> radius(points.size(), std::numeric_limits<double>::infinity());
    if (points.size() < 3) {
        return radius;
    }

    constexpr double epsilon = 1e-12;
    for (std::size_t index = 1; index + 1 < points.size(); ++index) {
        const Vec3 ab = points[index] - points[index - 1];
        const Vec3 bc = points[index + 1] - points[index];
        const Vec3 ac = points[index + 1] - points[index - 1];

        const double a = norm(ab);
        const double b = norm(bc);
        const double c = norm(ac);
        const double denom = a * b * c;
        if (denom <= epsilon) {
            continue;
        }

        const double twice_area = norm(cross(ab, bc));
        const double curvature = 2.0 * twice_area / denom;
        if (curvature > epsilon) {
            radius[index] = 1.0 / curvature;
        }
    }

    return radius;
}

[[nodiscard]] ConstraintReport compute_separation_report(
    const std::vector<Vec3>& points,
    const std::vector<double>& arclengths,
    double separation_world,
    int neighbor_exclusion
) {
    int compliant_count = 0;
    const double separation_sq = separation_world * separation_world;

    for (std::size_t i = 0; i < points.size(); ++i) {
        double nearest_allowed_sq = std::numeric_limits<double>::infinity();
        bool found_candidate = false;

        for (std::size_t j = 0; j < points.size(); ++j) {
            if (i == j) {
                continue;
            }
            if (std::abs(static_cast<long long>(j) - static_cast<long long>(i)) <= neighbor_exclusion) {
                continue;
            }
            if (std::abs(arclengths[j] - arclengths[i]) < separation_world) {
                continue;
            }

            found_candidate = true;
            nearest_allowed_sq = std::min(nearest_allowed_sq, squared_distance(points[i], points[j]));
        }

        if (!found_candidate || nearest_allowed_sq >= separation_sq) {
            ++compliant_count;
        }
    }

    return ConstraintReport{
        .compliant_count = compliant_count,
        .total_count = static_cast<int>(points.size()),
    };
}

[[nodiscard]] ConstraintReport compute_curvature_report(
    const std::vector<double>& radius,
    double min_radius_world
) {
    int compliant_count = 0;
    for (double value : radius) {
        if (value >= min_radius_world) {
            ++compliant_count;
        }
    }

    return ConstraintReport{
        .compliant_count = compliant_count,
        .total_count = static_cast<int>(radius.size()),
    };
}

}  // namespace

double ConstraintReport::ratio() const {
    if (total_count <= 0) {
        return 0.0;
    }
    return static_cast<double>(compliant_count) / static_cast<double>(total_count);
}

ValidationReport validate_polyline_constraints(
    const std::vector<Vec3>& points,
    double world_to_nm,
    double separation_nm,
    double min_curvature_radius_nm,
    int neighbor_exclusion
) {
    if (points.size() < 3) {
        throw std::invalid_argument("need at least 3 points");
    }
    if (world_to_nm <= 0.0) {
        throw std::invalid_argument("world_to_nm must be > 0");
    }

    const double separation_world = separation_nm / world_to_nm;
    const double min_radius_world = min_curvature_radius_nm / world_to_nm;
    const std::vector<double> arclengths = compute_arclengths(points);
    const std::vector<double> radius = compute_curvature_radius(points);

    return ValidationReport{
        .separation =
            compute_separation_report(points, arclengths, separation_world, neighbor_exclusion),
        .curvature = compute_curvature_report(radius, min_radius_world),
    };
}

}  // namespace autorigami

File: src/autorigami/validation.py
from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from scipy.spatial import KDTree


@dataclass(frozen=True)
class ConstraintReport:
    compliant_count: int
    total_count: int

    @property
    def ratio(self) -> float:
        if self.total_count <= 0:
            return 0.0
        return self.compliant_count / self.total_count


@dataclass(frozen=True)
class ValidationReport:
    separation: ConstraintReport
    curvature: ConstraintReport


def _curvature_radius(points: np.ndarray) -> np.ndarray:
    n = points.shape[0]
    radius = np.full(n, np.inf, dtype=float)
    if n < 3:
        return radius

    p0 = points[:-2]
    p1 = points[1:-1]
    p2 = points[2:]

    a = np.linalg.norm(p1 - p0, axis=1)
    b = np.linalg.norm(p2 - p1, axis=1)
    c = np.linalg.norm(p2 - p0, axis=1)

    cross = np.linalg.norm(np.cross(p1 - p0, p2 - p1), axis=1)
    kappa = np.zeros_like(a)
    denom = a * b * c
    valid = denom > 1e-12
    kappa[valid] = 2.0 * cross[valid] / denom[valid]

    with np.errstate(divide="ignore"):
        local_radius = np.where(kappa > 1e-12, 1.0 / kappa, np.inf)
    radius[1:-1] = local_radius
    return radius


def validate_polyline_constraints(
    points: np.ndarray,
    world_to_nm: float,
    separation_nm: float,
    min_curvature_radius_nm: float,
    neighbor_exclusion: int = 8,
) -> ValidationReport:
    pts = np.asarray(points, dtype=float)
    if pts.ndim != 2 or pts.shape[1] != 3:
        raise ValueError("points must have shape (N, 3)")
    if pts.shape[0] < 3:
        raise ValueError("need at least 3 points")
    if world_to_nm <= 0:
        raise ValueError("world_to_nm must be > 0")

    sep_world = separation_nm / world_to_nm
    min_radius_world = min_curvature_radius_nm / world_to_nm
    seg = np.linalg.norm(np.diff(pts, axis=0), axis=1)
    arclen = np.concatenate(([0.0], np.cumsum(seg)))

    tree = KDTree(pts)
    k = min(64, pts.shape[0])
    dists, idxs = tree.query(pts, k=k)
    dist_rows = np.asarray(dists, dtype=float)
    idx_rows = np.asarray(idxs, dtype=np.int64)

    separation_ok = np.zeros(pts.shape[0], dtype=bool)
    for i in range(pts.shape[0]):
        ok = True
        for d, j in zip(dist_rows[i], idx_rows[i]):
            idx = int(j)
            if i == idx:
                continue
            if abs(idx - i) <= neighbor_exclusion:
                continue
            if abs(float(arclen[idx] - arclen[i])) < sep_world:
                continue
            if float(d) < sep_world:
                ok = False
                break
            break
        separation_ok[i] = ok

    radius = _curvature_radius(pts)
    curvature_ok = radius >= min_radius_world

    return ValidationReport(
        separation=ConstraintReport(
            compliant_count=int(np.count_nonzero(separation_ok)),
            total_count=int(separation_ok.size),
        ),
        curvature=ConstraintReport(
            compliant_count=int(np.count_nonzero(curvature_ok)),
            total_count=int(curvature_ok.size),
        ),
    )
